core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java

```diff
@@ -925,34 +933,157 @@ public boolean isHashValid(String hash) {
+        }
+
+        private static byte[] generateSecretKey(PBEKeySpec spec)  {
+            SecretKeyFactory secretKeyFactory = null;
+            byte[] hash = new byte[0];
+            try {
+                secretKeyFactory = SecretKeyFactory.getInstance(PBKDF2_ALGORITHM);
+                hash = secretKeyFactory.generateSecret(spec).getEncoded();
+            } catch (NoSuchAlgorithmException e) {
+                LOGGER.log(WARNING, "No such algorithm found for encode",  e);
+            } catch (InvalidKeySpecException e) {
+                LOGGER.log(WARNING, "Invalid key spec exception",  e);
+            }
+            return hash;
+        }
+
+        private static String toHex(byte[] secretKey)  {
+
+            BigInteger bi = new BigInteger(1, secretKey);
+            String saltHashValue = bi.toString(RADIX_LENGTH);
+
+            int paddingLength = (secretKey.length * HEX_TWO) - saltHashValue.length();
+            if (paddingLength > ZERO) {
+                return String.format("%0" + paddingLength + "d", ZERO) + saltHashValue;
+            } else {
+                return saltHashValue;
+            }
+        }
+
+        @SuppressFBWarnings(value = {"DMI_RANDOM_USED_ONLY_ONCE", "PREDICTABLE_RANDOM"}, justification = "https://github.com/spotbugs/spotbugs/issues/1539 and doesn't need to be secure, we're just not hardcoding a 'wrong' password")
+        private static byte[] generateSalt() {
+            SecureRandom random = new SecureRandom();
+            byte[] salt = new byte[16];
+            random.nextBytes(salt);
+            return salt;
+        }
+
+        @Override
+        public boolean isHashValid(String hash)    {
+            Matcher matcher = PBKDF2_PATTERN.matcher(hash);
+            if (matcher.matches()) {
+                String logNumOfRound = matcher.group(1);
+                // no number format exception due to the expression
+                int logNumOfRoundInt = Integer.parseInt(logNumOfRound);
+                if (logNumOfRoundInt > 0 && logNumOfRoundInt <= MAXIMUM_PBKDF2_LOG_ROUND) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private static boolean validatePassword(String originalPassword, String storedPassword) {
+            String[] parts = storedPassword.split("[:$]");
+            int iterations = Integer.parseInt(parts[3]);
+
+            byte[] salt = Util.fromHexString(parts[4]);
+            byte[] hash = Util.fromHexString(parts[5]);
+
+            PBEKeySpec spec = new PBEKeySpec(originalPassword.toCharArray(),
+                    salt, iterations, hash.length * HASH_LENGTH);
+
+            byte[] generatedHashValue = generateSecretKey(spec);
+
+            int diff = hash.length ^ generatedHashValue.length;
+            for (int i = ZERO; i < hash.length && i < generatedHashValue.length; i++)
+            {
+                diff |= hash[i] ^ generatedHashValue[i];
+            }
```

Comment:
compare security tokens (byte[]) with [`MessageDigest.isEqual(byte[], byte[])`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/MessageDigest.html#isEqual(byte%5B%5D,byte%5B%5D)) rather than writing your own.

(will need an import)
```suggestion
            return MessageDigest.isEqual(hash, generatedHashValue);
```

Reply:
updated the same